Add Drupal 8 target

Kazanir opened the pull request:
Got it to work with Repo.Auth, yay!

fredemmott commented:
\o/

Please can you edit the README, in the same pattern to the others, explaining how this was created?

Kazanir commented:
Done. :)

rrh commented on targets/drupal8/Drupal8Target.php:
```diff
@@ -0,0 +1,58 @@
+      ->installDatabase();
+    $current_dir = __DIR__;
+    chdir($this->getSourceRoot());
+    shell_exec('find . -name *.html.twig | drush scr sites/default/setup.php');
```
Can you check the version/behavior of drush prior to calling it, so that incompatibilities can be flagged early?

Kazanir replied:
Yeah I can figure something out for checking this and then just throw up a warning about "Repo Auth is about to fail if you're using it" and skip the setup script.

rrh commented on README.md:
```diff
@@ -107,6 +107,26 @@ provided by the devel module. Default values were used, except:
+Drupal 8
+--------
+
+As above, aims to be realistic. Demo data is from the `devel_generate` module
```
Would somebody familiar with devel_generate and drupal8 be able to easily upgrade the contents of this target in 6 months to a more recent version of drupal, etc?  I, for one, have no experience with devel_generate, so would like more detailed instructions on how each of the companion files (db dumps, static content, etc, etc) were generated.

Kazanir replied:
Pretty easily -- there is a self-explanatory set of links and a UI for the module. That said, maybe the right way is not to include a DB dump at all, but rather to use Drush commands (including the batch that `devel_generate` provides) and install/generate the site from scratch as part of the installer?

rrh commented:
This will close issue https://github.com/hhvm/oss-performance/issues/42

Kazanir commented:
Ok, I added that check around the Drush version and a warning to the output if it fails. With that in mind and after talking to Fred about using Drush for live provisioning (verdict: no) I think we're good to go. I foresee being involved with Drupal and a fan of HHVM for quite a while so I'll try to update regularly when Drupal 8 gets to RCs or full releases. :)

fredemmott commented:
Currently buggy:

```
\nFatal error: Class undefined: __TwigTemplate_7e866b99f6b45929e048f6069cf00c5ccd95a8dfb40d0aaacbeb87e6c935afb7 in /tmp/hhvm-nginxYG6YGd/drupal-8.0.0-beta10/core/lib/Drupal/Core/Template/TwigEnvironment.php on line 141
```

Todo:
 - always run the twig setup script (so that repo auth is basically an identical path, apart from building the repo)
 - if drush is needed, bundle it.

Kazanir commented:
Doh. The settings changes I made also assumed that the pre-provisioning of templates was done. With the bundled Drush (which is now always used for consistency) this script should always run. (It throws a few warnings which I'll track down but they seem to be immaterial for the time being.)

rrh commented:
I cloned https://github.com/kazanir/oss-performance and ran it.  I noticed several errors in one of the log files (perhaps stdout? stderr? from the oss-performance hack driver script).

include_once(/var/www/d8test.ph/html/core/themes/bartik/bartik.theme)failed to open stream: No such file or directory Extension.php:145

Invalid argument supplied for foreach() twig.engine:160
constant(): Couldn't find constant UPDATE_NOT_SECURE Core.php:1463  
constant(): Couldn't find constant UPDATE_REVOKED Core.php:1463      
constant(): Couldn't find constant UPDATE_NOT_SUPPORTED Core.php:1463
constant(): Couldn't find constant UPDATE_NOT_CURRENT Core.php:1463  
constant(): Couldn't find constant UPDATE_CURRENT Core.php:1463      
constant(): Couldn't find constant UPDATE_NOT_SECURE Core.php:1463   
constant(): Couldn't find constant UPDATE_REVOKED Core.php:1463      
constant(): Couldn't find constant UPDATE_NOT_SUPPORTED Core.php:1463^

rrh commented:
Can you please document the inflated site's admin login and password?  I tried "root" "root", but didn't get admin access.

Kazanir commented:
Yeah, those are Drush reporting to stderr. I'll get them fixed upstream but the benchmarks ran in spite of them. The first line is strange though; it's a cached file path from the original install...I can get that out of the database and I'll switch the username and password to root/root also. (I think I used admin/test but can't remember for sure.)

jwatzman commented:
Another warning... requires the `php5-mysql` package to be installed, otherwise you get a strange failure mode:

```
$ hhvm perf.php  --drupal8 --hhvm=$HOME/bin/hhvm

**********************************************************
** 2015-05-28 15:14:19 PDT
** Configuration: Drupal8Target on /home/jwatzman/bin/hhvm

**************************
** 2015-05-28 15:14:19 PDT
** Installing framework
PHP Fatal error:  Undefined class constant 'MYSQL_ATTR_USE_BUFFERED_QUERY' in /tmp/hhvm-nginx7N8isG/drupal-8.0.0-beta10/core/lib/Drupal/Core/Database/Driver/mysql/Connection.php on line 78
PHP Fatal error:  Uncaught exception 'Symfony\Component\DependencyInjection\Exception\ServiceCircularReferenceException' with message 'Circular reference detected for service "database", path: "module_handler -> cache.bootstrap -> cache.backend.database -> database -> string_translation -> state -> keyvalue.database".' in /tmp/hhvm-nginx7N8isG/drupal-8.0.0-beta10/core/vendor/symfony/dependency-injection/Symfony/Component/DependencyInjection/Container.php:296
Stack trace:
#0 /tmp/hhvm-nginx7N8isG/drupal-8.0.0-beta10/sites/default/files/php/service_container/service_container_prod/060740ad3fb874289b728d66f52a7e7603a1ac52921ab4ccc06bc9d1de5308ee.php(3837): Symfony\Component\DependencyInjection\Container->get('database')
#1 /tmp/hhvm-nginx7N8isG/drupal-8.0.0-beta10/core/vendor/symfony/dependency-injection/Symfony/Component/DependencyInjection/Container.php(326): Drupal\Core\DependencyInjection\Container\prod\service_container_prod->getKeyvalue_DatabaseService()
#2 /tmp/hhvm-nginx7N8isG/drupal-8.0.0-beta10/core/lib/Drupal/Core/KeyV in /tmp/hhvm-nginx7N8isG/drupal-8.0.0-beta10/core/vendor/symfony/dependency-injection/Symfony/Component/DependencyInjection/Container.php on line 296

**************************
** 2015-05-28 15:14:22 PDT
** There is no setUpTest

**************************
** 2015-05-28 15:14:22 PDT
** Starting Nginx
nginx: [alert] could not open error log file: open() "/var/log/nginx/error.log" failed (13: Permission denied)

**************************
** 2015-05-28 15:14:22 PDT
** Starting PHP Engine
mapping self...
mapping self took 0'00" (1282 us) wall time
loading static content...
loading static content took 0'00" (0 us) wall time

**************************
** 2015-05-28 15:14:24 PDT
** Running sanity check

Warning: Failed to open http://jwatzman-trashcan:8090/ () in /home/jwatzman/oss-performance-drupal8/base/PerfTarget.php on line 28

Fatal error: Uncaught exception 'HH\InvariantException' with message 'Failed to find string "Read more" in http://jwatzman-trashcan:8090/' in :
Stack trace:
#0 /home/jwatzman/oss-performance-drupal8/base/PerfTarget.php(33): HH\invariant_violation()
#1 /home/jwatzman/oss-performance-drupal8/base/PerfRunner.php(101): PerfTarget->sanityCheck()
#2 /home/jwatzman/oss-performance-drupal8/base/PerfRunner.php(46): PerfRunner::RunWithOptionsAndEngine()
#3 /home/jwatzman/oss-performance-drupal8/base/PerfRunner.php(19): PerfRunner::RunWithOptions()
#4 /home/jwatzman/oss-performance-drupal8/perf.php(15): PerfRunner::RunWithArgv()
#5 /home/jwatzman/oss-performance-drupal8/perf.php(19): perf_main()
#6 {main}
```

fredemmott commented:
- if all of that stuff is expected, please redirect stderr from drush
- there's some disagreement on if cached or uncached is 'correct' - so could you change this to add both --drupal8-cached and --drupal8-uncached? SugarCRM is an existing example of two targets from the same source

Kazanir commented:
Thanks for the feedback guys. I will be able to get back to this on the weekend or early next week, here's the plan:

- [ ] Add a page-cache-on and page-cache-off target
- [ ] Get those errors fixed upstream or swallow them if I can confirm they are red herrings
- [ ] Change the admin creds to root/root
- [ ] Fix any cached crap related to my original install
- [ ] Change Drush to run via HHVM; this will fix the PHP5/MySQL error

Stay tuned.

rrh commented:
Thanks for doing this!

Kazanir commented:
Ok, I think all this is done...I'm going to add a README to the Drupal 8 target folder attempting to document how I put the thing together, as well, and then give it a bunch more testing tonight.

Kazanir commented:
Looks like the current PHP7 nightly is a bit broken but this works with a version from a few days ago. To get PHP5 to work, the `opcache.save_comments` setting has to be 1 (for the same reasons as with Doctrine and any other framework (e.g. ZF2) that relies on its annotation parsing.) Not sure what to do about that other than document the needed change? According to PHP's manual that setting cannot be changed at runtime.

fredemmott commented:
If an ini change is required, it should generate that ini change :( might need some more plumbing.

Kazanir commented:
Ok. Made the small script fixes requested, added notes about packing up Drush, and fixed the D8 target URLs (Drupal 8 has removed ?q= in favor of clean URLs, I guess.)

I don't know how to handle the .ini setting change in a way that is target-specific.

fredemmott commented:
What's the downside to always having that option set?

rrh commented:
I reran the latest commit in this fork and it looked fine under HHVM.  However, running a fairly reliable private build of php5.6 cgi (no-zts), I get these errors in the nginx-error.log file (the line breaks are my own doing):

2015/06/02 17:52:19 [error] 17071#0: *5472 
FastCGI sent in stderr: "Uncaught PHP Exception
Doctrine\Common\Annotations\AnnotationException: " [Semantical Error]
The class "Drupal\Core\Display\Annotation\PageDisplayVariant" is not
annotated with @Annotation. Are you sure this class can be used as
annotation? If so, then you need to add @Annotation to the _class_
doc comment of "Drupal\Core\Display\Annotation\PageDisplayVariant".
If it is indeed no annotation, then you need to add
@IgnoreAnnotation("PageDisplayVariant") to the _class_ doc comment
of class Drupal\block\Plugin\DisplayVariant\BlockPageVariant." at 
/tmp/perf/drupal-8.0.0-beta11/core/vendor/doctrine/annotations/lib/Doctr
ine/Common/Annotations/AnnotationException.php line 54" while reading
response header from upstream, client: 127.0.0.1, server: , request:
"GET /node/180 HTTP/1.0", upstream: "fastcgi://127.0.0.1:8092", host:
"ubuntu:8090"

A page load through the browser (not through siege) says:
Drupal 8 OSS Performance Test (No Page Cache)
Error
The website encountered an unexpected error. Please try again later.

Kazanir commented:
Yes, that's the ini setting. The conf/php.ini has `opcache.save_comments=0` which means that comments aren't saved by OpCache and thus Doctrine-style annotations don't work. The PHP manual documents this problem here:

http://php.net/manual/en/opcache.configuration.php#ini.opcache.save-comments

The only effect leaving `opcache.save_comments=1` would have (I think) is increasing the size of the opcode cache. I just assumed someone had turned it off in the first place for a reason... :)

rrh commented:
aha.  thanks.  I should have read the history here more closely.  I manually change the save comments to a 1, and php7 now seems happy.  I agree with one of the other commentators here that this change should be done as needed by the driver script, so this can work out of the box.
thanks again for doing this

Kazanir commented:
I guess we could just shell out during setup and fix it back during teardown, like so:

```
sed -i 's/opcache\.save_comments=0/opcache\.save_comments=1/' conf/php.ini
```

Thoughts?

Kazanir commented:
Oh there is no actual teardown function, only a CLI option for a tearDownTest. It's up to Fred if he thinks adding a `cleanup()` method to the PerfTarget (which would run at the end of any given run) is a good idea; it looks like our options are that or just changing `save_comments=1` (which is the default anyway.)

fredemmott commented:
going to re-run a few without that option and see how results compare

fredemmott commented:
With page cache and opcache.save_comments=0 removed:

HHVM: `        "Nginx avg bytes": 73434.397078145,`
PHP5: `        "Nginx avg bytes": 38218.757210582,`
PHP7: `        "Nginx avg bytes": 38263.393704441,`

Can't land this until those are more similar

Kazanir commented:
I updated the database dumps, URL, and README to be more the-next-guy friendly, and rebased on top of Fred's latest commits. Leaving it for him to squash when he merges.
